ast.ts

```diff
@@ -7,12 +7,27 @@ export type Type =
   | {tag: "none"}
   | {tag: "class", name: string}
   | {tag: "either", left: Type, right: Type }
+  // generator type
+  | { 
+    tag: "generator", 
+    type: Type // type of item generated each time calling next()
+  }
 
 export type Parameter<A> = { name: string, type: Type }
 
 export type Program<A> = { a?: A, funs: Array<FunDef<A>>, inits: Array<VarInit<A>>, classes: Array<Class<A>>, stmts: Array<Stmt<A>> }
 
 export type Class<A> = { a?: A, name: string, fields: Array<VarInit<A>>, methods: Array<FunDef<A>>}
+  | Iterable<A> // iterable interface
+
+// we hope that list/set/tuple/dict/string will implement (i.e. inherit) 
+// this Iterable interface, and provide implementations of next() and has_next(), 
+// so that we may parse comprehension expressions to IR by calling them
+export type Iterable<A> = 
+  {
+    a?: A,
+    type: Type // type of iterable object
+  }
```

Comment:
I don't know that I understand how this type works. Can you say more about it?

For example, if a user defines a new class that has `next` and `hasnext` methods, would instances of it subtype to `Iterable` somehow?

I'm not saying that has to be implemented right away. But I don't understand from this document how `Iterable` gets used – where in the type-checker will `Iterable`s be constructed, for instance?

Reply:
Yeah I think this is one thing we are still trying to make design decisions on. We are thinking of having an interface ```Iterable``` that supports ```next()``` and ```has_next()``` so that our comprehensions know how to traverse through an iterable object such as a list. In the current state of the implementation (and also in Python), an approximation of an interface is an abstract class.

We imagine that, say, the list group will implement ```List``` as a built-in class, and they need to inherit the abstract class ```Iterable``` and implement the above methods. For any user-defined type that needs to support iteration, we force them to inherit the ```Iterable``` class too.

Alternatively, the ```List``` class doesn't have to inherit the abstract class ```Iterable```. Instead, whenever we see a ```List``` type after the ```in``` keyword in a comprehension, we construct an ```Iterable``` object that has the above two methods (and the ```type``` field in the above AST would be ```List```).

Under the hood, we will transform a comprehension expression ```[val for val in <iterable>]``` to something like:
```python
res: List[<type>] = []
val: <type> = None
while <iterable>.has_next():
    val = <iterable>.next()
    #yield val # Oh, it seems that we need to add a yield statement in AST/IR as well
    res.append(val)
```

We know there are still a few issues with the above approaches, such as how to re-iterate through an object if the inheritance approach is adopted, and how to implement a built-in function ```next(g: generator)```.